### ISSUE TITLE
sentencepiece protobuf mismatch (only with Horovod!)

### ISSUE THREAD

**Issue by jchwenger** [horovod/horovod#1906]
**Environment:**
Horovod version: 0.19.1,
mpirun (Open MPI): 4.0.3
CUDA Version 10.1.243
nccl: libnccl2=2.4.8-1+cuda10.1 libnccl-dev=2.4.8-1+cuda10.10.1
gcc (Ubuntu 7.4.0-1ubuntu1~18.04.1) 7.4.0
(installed using [the recommended steps](https://github.com/horovod/horovod#install))
sentencepiece 0.1.86 (compiled from source) 
and also the python packages:
sentencepiece        0.1.85             
tf-sentencepiece     0.1.85  
libprotoc 3.11.4

**Checklist:**
1. Did you search issues to find if somebody asked this question before?

This is definitely close to [this question](https://github.com/horovod/horovod/issues/1492), and I have also seen [this](https://github.com/google/sentencepiece/issues/383) and one [workaround](https://github.com/google/sentencepiece/issues/325#issuecomment-546789546). I should not that this is related to tensorflow serving, whereas here it's TF 1.14.

**Bug report:**

Hi,

I'm facing an odd one. I wish to train a gpt-2-like model using the sentencepiece encoder, which my machine is currently doing. I can
1. train using openAI's encoder, using their English vocab table, including on two GPUs with Horovod (Yay!).
2. Train *on a single GPU* with sentencepiece without Horovod.
3. However, when trying to use horovod with the sentencepiece encoder, I get this:

```
[1,1]<stderr>:[libprotobuf FATAL external/protobuf_archive/src/google/protobuf/stubs/common.cc:86] This program was compiled against version 3.6.1 of the Protocol Buffer runtime library, which is not compatible with the installed version (3.7.1).  Contact the program author for an update.  If you compiled the program yourself, make sure that your headers are from the same version of Protocol Buffers as your link-time library.  (Version verification failed in "/sentencepiece/src/builtin_pb/sentencepiece_model.pb.cc".)
[1,0]<stderr>:[libprotobuf FATAL external/protobuf_archive/src/google/protobuf/stubs/common.cc:86] This program was compiled against version 3.6.1 of the Protocol Buffer runtime library, which is not compatible with the installed version (3.7.1).  Contact the program author for an update.  If you compiled the program yourself, make sure that your headers are from the same version of Protocol Buffers as your link-time library.  (Version verification failed in "/sentencepiece/src/builtin_pb/sentencepiece_model.pb.cc".)
[1,1]<stderr>:terminate called after throwing an instance of 'google::protobuf::FatalException'
[1,1]<stderr>:  what():  This program was compiled against version 3.6.1 of the Protocol Buffer runtime library, which is not compatible with the installed version (3.7.1).  Contact the program author for an update.  If you compiled the program yourself, make sure that your headers are from the same version of Protocol Buffers as your link-time library.  (Version verification failed in "/sentencepiece/src/builtin_pb/sentencepiece_model.pb.cc".)
```

As mentioned above, I have seen people talking about it, and am still looking to see if I can apply their solutions to my problem. They could not make Horovod or Tensorflow to work at all, it seems, which is not my case.

The installed version of protobuf, is also far higher:
```
$ conda list | grep protobuf
libprotobuf               3.11.4               hd408876_0  
protobuf                  3.11.4           py37he6710b0_0  
$ protoc --version
libprotoc 3.11.4
```
I saw that issue rear its incompatible head when people were trying to compile TF from source, which I'd rather avoid for now if I can. The really odd thing is that I had no problem building horovod, and am able to train with it as long as sentencepiece is not part of the equation, or making TF 1.14 work with sentencenpiece on a single GPU. 

Thanks a lot for any hint!

**Comment by tgaddair** [comment]
Hey @jchwenger, how did you install / build sentencepiece?  Looks like it gets installed with its own version of protobuf (which I'm guessing was 3.6.1).  TensorFlow comes with its own version of protobuf that it uses, which for TensorFlow 1.14 is Protobuf 3.7.1 (see: [here](https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/workspace.bzl#L374)), so that's where 3.7.1 is coming from.

My suggestion would be to try and rebuild sentencepiece with Protobuf 3.7.1 as described in the linked workaround.  Alternatively, you may want to try downgrading to TensorFlow 1.13.1 and reinstalling Horovod, which would use Protobuf 3.6.1.

**Comment by jchwenger** [comment]
Hi @tgaddair, thanks, I'll try that!

**Comment by jchwenger** [comment]
Hi @tgaddair, I'm sorry my previous response was quite inadequate, I wanted to send something quickly. So I compiled sentencepiece [from source](https://github.com/google/sentencepiece#c-from-source) and it's 0.1.86, with the two pip packages at version 0.1.85. I can locate the "300600" relating to protobuf, so indeed making the versions compatible seems the way to go.

 I'm looking into the whole process of compiling protobuf from source as well, and digging into how one can modify the `sentencepiece_processor` inside tensorflow (as per the [workaround mentioned](https://github.com/google/sentencepiece/issues/325#issuecomment-546789546)), but what I can't wrap my head around is that with just tf 1.14 and sentencepiece as is things do work, I can train my model on one gpu ([here](https://github.com/jchwenger/gpt-2/blob/titan/train.py)), but when I run the other script with Horovod, which also works on multiple gpus without sentencepiece ([here](https://github.com/jchwenger/gpt-2/blob/titan/train-horovod.py), I get the error. And I did not compile tf from source nor encountered any problem installing openmpi and horovod either.
